### PR TITLE
pageserver: post-shard-split layer rewrites (2/2)

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

- After a shard split of a large existing tenant, child tenants can end up with oversized historic layers indefinitely, if those layers are prevented from being GC'd by branchpoints.

This PR follows https://github.com/neondatabase/neon/pull/7531, and adds rewriting of layers that contain a mixture of needed & un-needed contents, in addition to dropping un-needed layers.

Closes: https://github.com/neondatabase/neon/issues/7504

## Summary of changes

- Add methods to ImageLayer for reading back existing layers
- Extend `compact_shard_ancestors` to rewrite layer files that contain a mixture of keys that we want and keys we do not, if unwanted keys are the majority of those in the file.
- Amend initialization code to handle multiple layers with the same LayerName properly
- Get rid of of renaming bad layer files to `.old` since that's now expected on restarts during rewrites.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: pageserver/src/tenant/storage_layer/image_layer.rs [hunk 1]
```diff
@@ -516,16 +520,36 @@ impl ImageLayerInner {
                 let key = Key::from_slice(&raw_key[..KEY_SIZE]);
                 assert!(key >= range.start);
 
+                let drop = !filter(&key);
```
Comment (VladLazar) [inline pageserver/src/tenant/storage_layer/image_layer.rs:523]: I think the code below is correct, but I find it hard to reason about.
You could do the following instead:
```
let flag = {
  if !filter(&key) {
    BlobFlag::Ignore
  } else {
    BlobFlag::None
  }
}
```

Pass this flag into `planner.handle` and don't modify the rest of the code.
This offloads the work to the planer and remove s the need for separate tests.

File: pageserver/src/tenant/storage_layer/image_layer.rs [hunk 2]
```diff
@@ -534,6 +558,52 @@ impl ImageLayerInner {
         Ok(planner.finish())
     }
 
+    /// Given a key range and filter function, generate a read plan and execute it, then for
+    /// each value we read write it onwards into the writer.
+    pub(super) async fn do_reads_and_write<F>(
```
Comment (VladLazar) [inline pageserver/src/tenant/storage_layer/image_layer.rs:563]: nit: Can we give this a different name since it's part of the public interface? Something like `rewrite_layer` or `filter_layer`.

Comment (VladLazar) [inline pageserver/src/tenant/storage_layer/image_layer.rs:563]: A test for this would be nice. You could look the tests in `delta_layer.rs` for the harness.

File: pageserver/src/tenant/timeline/compaction.rs [hunk 3]
```diff
@@ -267,9 +269,7 @@ impl Timeline {
             // shard but an old generation, or they belonged to another shard.  This also implicitly
             // guarantees that the layer is persistent in remote storage (as only remote persistent
             // layers are carried across shard splits, any local-only layer would be in the current generation)
-            if layer.metadata().generation == self.generation
-                && layer.metadata().shard.shard_count == self.shard_identity.count
-            {
+            if layer.metadata().generation == self.generation {
```
Comment (VladLazar) [inline pageserver/src/tenant/timeline/compaction.rs:272]: nit: can you update the comment above as well?

**Review by VladLazar — COMMENTED** [review]
Some comments, but generally looks good.

For my understanding: commit `pageserver: init should not assume local LayerNames are unique` is required for cases where the compaction task panics mid rewrites?

File: pageserver/src/tenant/timeline/init.rs [hunk 4]
```diff
@@ -99,14 +98,12 @@ impl LocalLayerFileMetadata {
 pub(super) enum Decision {
     /// The layer is not present locally.
     Evicted(LayerFileMetadata),
-    /// The layer is present locally, but local metadata does not match remote; we must
-    /// delete it and treat it as evicted.
-    UseRemote {
+    /// The layer is present locally, and metadata matches: we may hook up this layer to the
+    /// existing file in local storage.
+    UseLocal {
```
Comment (koivunej) [inline pageserver/src/tenant/timeline/init.rs:103]: Not sure if I understand what does this comment mean, or why is UseRemote removed.

File: pageserver/src/tenant/timeline/compaction.rs [hunk 5]
```diff
@@ -303,6 +363,10 @@ impl Timeline {
             remote_client.wait_completion().await?;
         }
 
+        fail::fail_point!("compact-shard-ancestors-persistent-kill", |_| {
+            std::process::abort()
+        });
```
Comment (koivunej) [inline pageserver/src/tenant/timeline/compaction.rs:368]: We already have the `exit` action.

File: pageserver/src/tenant/storage_layer/layer.rs [hunk 6]
```diff
@@ -1843,6 +1842,27 @@ impl ResidentLayer {
         }
     }
 
+    /// Read all they keys in this layer in the given range, and write them all to
+    /// the provided writer.  Return the number of keys written.
+    #[tracing::instrument(level = tracing::Level::DEBUG, skip_all, fields(layer=%self))]
+    pub(crate) async fn rewrite<'a, F>(
+        &'a self,
+        range: Range<Key>,
+        filter: F,
```
Comment (koivunej) [inline pageserver/src/tenant/storage_layer/layer.rs:1851]: Do we have some need for different kind of filters?

Ok, it's for not hoisting the shard key selection up here. Unsure do we need these low level ops to be generic, but here I guess it doesn't hurt here.

Rewrite could be mean anything and the naming direction is at least different than of mine with `copy_lsn_prefix`, which is however quite similar to what this is doing, the generic filter fn for `copy_delta_prefix` is just `|_k, lsn| lsn <= until`, but I wanted to do it non-generic.

File: pageserver/src/tenant/timeline/init.rs
Comment (koivunej) [inline pageserver/src/tenant/timeline/init.rs:1]: I do not understand the changes in this file, are they related?

File: pageserver/src/tenant/timeline/layer_manager.rs [hunk 7]
```diff
@@ -212,13 +212,35 @@ impl LayerManager {
...
-        assert!(rewrite_layers.is_empty());
+            Self::delete_historic_layer(old_layer, &mut updates, &mut self.layer_fmgr);
 
+            Self::insert_historic_layer(
+                new_layer.as_ref().clone(),
+                &mut updates,
+                &mut self.layer_fmgr,
+            );
+
+            metrics.resident_physical_size_sub(old_layer.layer_desc().file_size);
```
Comment (koivunej) [inline pageserver/src/tenant/timeline/layer_manager.rs:241]: Wouldn't this double count with the substractions done by the layer on drop? Please only use the `record_new_file_metrics`.

File: test_runner/regress/test_sharding.py [hunk 8]
```diff
@@ -177,7 +177,16 @@ def test_sharding_split_unsharded(
...
 
 
-def test_sharding_split_compaction(neon_env_builder: NeonEnvBuilder):
+@pytest.mark.parametrize(
+    "failpoint",
+    [
+        None,
+        "compact-shard-ancestors-localonly-kill",
+        "compact-shard-ancestors-enqueued-kill",
+        "compact-shard-ancestors-persistent-kill",
```
Comment (koivunej) [inline test_runner/regress/test_sharding.py:186]: We don't need to name the failpoint for the action. Plain `return` would fail for these failpoints when defined with `fail!(name)`.

File: pageserver/src/tenant/timeline/init.rs [hunk 9]
```diff
@@ -201,25 +203,15 @@ pub(super) fn cleanup(path: &Utf8Path, kind: &str) -> anyhow::Result<()> {
     std::fs::remove_file(path).with_context(|| format!("failed to remove {kind} at {path}"))
 }
 
-pub(super) fn cleanup_local_file_for_remote(
-    local: &LocalLayerFileMetadata,
-    remote: &LayerFileMetadata,
-) -> anyhow::Result<()> {
-    let local_size = local.metadata.file_size();
-    let remote_size = remote.file_size();
+pub(super) fn cleanup_local_file_for_remote(local: &LocalLayerFileMetadata) -> anyhow::Result<()> {
```
Comment (koivunej) [inline pageserver/src/tenant/timeline/init.rs:206]: Is this still the case or not, the generation numbers are now part of the file identity, so why is this not a case of "local only"?

File: pageserver/src/tenant/storage_layer/image_layer.rs (hunk 2 above)
Comment (koivunej) [reply pageserver/src/tenant/storage_layer/image_layer.rs:563]: Hmm as far as unifying with my work goes, I kinda like the `filter_layer` or just `filter` (because we are already calling it on a layer) makes sense.

Comment (koivunej) [inline pageserver/src/tenant/storage_layer/image_layer.rs:563]: For the delta copy lsn prefix I did not reuse the existing vectored reads system completely because in the case that majority of the file needs to be read, we are unnecessarily building up those planned reads where as we could just process them as soon as they are built.

I guess you were thinking that you do the size estimate already, so that is not needed here? But then it would for me suggest that we shouldn't make this generic over the filter function.

File: pageserver/src/tenant/timeline/compaction.rs [hunk 10]
```diff
@@ -251,9 +253,9 @@ impl Timeline {
 
             // Don't bother re-writing a layer if it is within the PITR window: it will age-out eventually
             // without incurring the I/O cost of a rewrite.
-            if layer_desc.get_lsn_range().end >= pitr_cutoff {
-                debug!(%layer, "Skipping rewrite of layer still in PITR window ({} >= {})",
-                    layer_desc.get_lsn_range().end, pitr_cutoff);
+            if layer_desc.get_lsn_range().end >= last_gc_cutoff {
+                debug!(%layer, "Skipping rewrite of layer still in GC window ({} >= {})",
+                    layer_desc.get_lsn_range().end, last_gc_cutoff);
```
Comment (koivunej) [inline pageserver/src/tenant/timeline/compaction.rs:258]: Is this not the wrong way around?

As in the layers that are "alive in the future" are within the cutoff. Or am I again thinking this the wrong way around?

File: pageserver/src/tenant/timeline/compaction.rs [hunk 11]
```diff
@@ -282,18 +282,78 @@ impl Timeline {
...
+            //   with it. If they are trying to promote from remote storage, that will fail, but this is the same
+            //   as for compaction generally: compaction is allowed to delete layers that readers might be trying to use.
+            // - We do not run concurrently with other kinds of compaction, so the only layer map writes we race with are:
+            //    - GC, which at worst witnesses us "undelete" a layer that they just deleted.
+            //    - ingestion, which only inserts layers, therefore cannot collide with us.
+            let resident = layer.download_and_keep_resident().await?;
+
+            let keys_written = resident
+                .rewrite(
+                    layer.layer_desc().key_range.clone(),
```
Comment (koivunej) [inline pageserver/src/tenant/timeline/compaction.rs:322]: I don't understand why is this range passed in. The layer knows it, it's in its name and summary.

**Review by koivunej — CHANGES_REQUESTED** [review]
At least we need to replace the `kill` action with `exit` and not add any code for it. I think we should try to align the design and impl with the delta copy lsn prefix work. I had a lot of things I couldn't understand, will check later today.

File: pageserver/src/tenant/timeline/compaction.rs (hunk 10 above)
Comment (jcsp) [reply pageserver/src/tenant/timeline/compaction.rs:258]: The idea is to rewrite only old layers: more recent layers are allowed to continue existing, because they'll likely get cleaned up anyway during GC.  This is to avoid doing a full rewrite of all the layers when we split a tenant: we instead tolerate some size bloat until the layers reach the GC cutoff, then if they _still_ exist, we interpret that as "this layer is probably pinned by some branch point, so rather than keep it forever in the unfiltered state, let's filter it"

File: pageserver/src/tenant/timeline/compaction.rs (hunk 3 above)
Comment (jcsp) [reply pageserver/src/tenant/timeline/compaction.rs:272]: done.

File: pageserver/src/tenant/timeline/init.rs (hunk 4 above)
Comment (jcsp) [reply pageserver/src/tenant/timeline/init.rs:103]: Previously, we had UseLocal, which meant "use remote metadata, and the local layer exists", and UseRemote, whichc meant "use remote metadata, and the local layer exists but doesn't match and should be deleted".

In this PR I've changed Decision to just reflect whether the layer is resident or not, and for the case where we want to delete the local file, I've moved that to DismissedLayer::BadMetadata.  This simplifies the logic, because a Decision now always refers to a layer that exists in remote storage.

Comment (jcsp) [reply pageserver/src/tenant/timeline/init.rs:103]: The naming was a bit weird though: I've changed this to call the variants Evicted and Resident instead of Evicted and UseLocal.

File: pageserver/src/tenant/timeline/compaction.rs (hunk 5 above)
Comment (jcsp) [reply pageserver/src/tenant/timeline/compaction.rs:368]: Huh, how did I miss that?  changed these to use 'exit' in f28f0ce36

File: pageserver/src/tenant/timeline/init.rs (hunk 9 above)
Comment (jcsp) [reply pageserver/src/tenant/timeline/init.rs:206]: Right now, we don't parse the generation number from the local path: it's just used opaquely to generate a more-unique path.

We could do that to disambiguate a bad file from a different generation (if the local file is new enough to have a generation), although currently the action we take is the same :thinking:

File: pageserver/src/tenant/storage_layer/image_layer.rs (hunk 1 above)
Comment (jcsp) [reply pageserver/src/tenant/storage_layer/image_layer.rs:523]: ah, a very nice suggestion.  c54ca636995b18fbe04425f0fff35cb030e14827

**Comment by jcsp** [comment]
>For my understanding: commit pageserver: init should not assume local LayerNames are unique is required for cases where the compaction task panics mid rewrites?

Right: where we have two layers on disk that have the same LayerName but different generation suffixes.

File: pageserver/src/tenant/timeline/compaction.rs (hunk 11 above)
Comment (jcsp) [reply pageserver/src/tenant/timeline/compaction.rs:322]: True -- the ResidentLayer doesn't know its own range but the Layer does.  Moved this down one layer in c54ca636995b18fbe04425f0fff35cb030e14827 to have Layer use its own desc.

File: pageserver/src/tenant/storage_layer/layer.rs (hunk 6 above)
Comment (jcsp) [reply pageserver/src/tenant/storage_layer/layer.rs:1851]: I could imagine using this for something arbitrary like breaking up some oversized layer into smaller ones, but for the moment the ShardIdentity is the only kind of filtering we care about.  I've made this a bit less genericc to just take a ShardIdentity instead of an arbitrary function

File: pageserver/src/tenant/timeline/layer_manager.rs (hunk 7 above)
Comment (jcsp) [reply pageserver/src/tenant/timeline/layer_manager.rs:241]: you're right -- this made sense in an earlier version (before generation numbers in paths) where layers had their physical removal disabled, but is wrong now.  Removed in 35690b525

File: pageserver/src/tenant/timeline/compaction.rs (hunk 10 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline/compaction.rs:258]: Aah ok so my point was correct but that is wanted behaviour. Ok makes sense.

How to stop me from thinking this is a bug in future :) You provided good rationale, however, this is racy. We compact (30/h) and refresh gc info (6/h) much more often than we do, however the `Timeline::latest_gc_cutoff_lsn` is advanced at once per hour. If last_gc_cutoff is `latest_gc_cutoff_lsn` this should be working as expected.

Idea for documenting the why so I don't report this as a bug when I go over it in editor:

```suggestion
            if layer_desc.get_lsn_range().end >= last_gc_cutoff {
                // intentionally only rewrite layers eligible for gc which must be
                // retained for branches
                debug!(%layer, "Skipping rewrite of layer still in GC window ({} >= {})",
                    layer_desc.get_lsn_range().end, last_gc_cutoff);
```

However ... This does make me think this is another case of possible gc and compaction problem? If we rewrite a layer which is doomed to be gc'd, removing it would be a panic?

I still haven't gotten to writing that test case.

File: pageserver/src/tenant/timeline/compaction.rs (hunk 11 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline/compaction.rs]: This may have been unnecessary, the `owner: &Arc<LayerInner>` was most likely already passed into this method?

But still, I question why is it needed in the first place.

File: pageserver/src/tenant/storage_layer/layer.rs [hunk 12]
```diff
@@ -1817,6 +1816,25 @@ impl ResidentLayer {
...
+    ) -> anyhow::Result<usize> {
+        use LayerKind::*;
+
+        let key_range = self.layer_desc().key_range.clone();
+
+        match self.downloaded.get(&self.owner.0, ctx).await? {
+            Delta(_) => anyhow::bail!(format!("cannot load_image_keys on a delta layer {self}")),
+            Image(i) => i.filter(key_range, shard_identity, writer, ctx).await,
+        }
+    }
```
Comment (koivunej) [inline pageserver/src/tenant/storage_layer/layer.rs:1836]: Additional reason why I added the copy_lsn_prefix into ResidentLayer (should had documented this) is to force the `download_and_keep_resident`. Not forcing the caller to do this will create an avenue for easily doing the download while holding the read lock on layers, which we currently have on (and have always had) on L0 -> L1 compaction.

But `owner: &Arc<LayerInner>` was most likely passed to the previous place of this method, as I noted in my other comment.

File: pageserver/src/tenant/timeline/init.rs (hunk 9 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline/init.rs:203]: Ok, I was thinking it was implemented in some other fashion it ended up being implemented. Shame on me for not building the test case which I thought was so easy to build :)

In the implementation I had in mind, it would be a matter of noticing the other file has no generation and other local file has a generation... But that is not simple either. Perhaps this works, I should now re-read it.

File: pageserver/src/tenant/timeline/init.rs
Comment (koivunej) [reply pageserver/src/tenant/timeline/init.rs:1]: Other threads alrady include this.

File: pageserver/src/tenant/timeline/compaction.rs (hunk 10 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline/compaction.rs:258]: > I'll edit this soon upon confirming this does the read+write locking on Timeline::layers.

Yes, this exhibits the same pattern but the layer locking is now obscured under `Timeline::rewrite_layers` -- I think that is not great because it's not a free standing operation because of underlying implementation, I'd just prefer to see it inlined. I think this is more risky than current legacy compaction because we explicitly target the gc candidates.

The risk would be mitigated by using AND keeping alive the `Timeline::latest_gc_cutoff_lsn.read()` until done.

File: pageserver/src/tenant/timeline/compaction.rs [hunk 13]
```diff
@@ -186,13 +186,15 @@ impl Timeline {
...
         let mut drop_layers = Vec::new();
-        let layers_to_rewrite: Vec<Layer> = Vec::new();
+        let mut layers_to_rewrite: Vec<Layer> = Vec::new();
 
-        // We will use the PITR cutoff as a condition for rewriting layers.
-        let pitr_cutoff = self.gc_info.read().unwrap().cutoffs.pitr;
+        // We will use the Lsn cutoff of the last GC as a threshold for rewriting layers: if a
+        // layer is behind this Lsn, it indicates that the layer is being retained beyond the
+        // pitr_interval, for example because a branchpoint references it.
+        let last_gc_cutoff = self.gc_info.read().unwrap().cutoffs.pitr;
```
Comment (koivunej) [inline pageserver/src/tenant/timeline/compaction.rs:197]: As noted in https://github.com/neondatabase/neon/pull/7531/files#r1602860120 and after, instead let's use `Timeline::latest_gc_cutoff_lsn.read()` guard for the whole operation, which will stall gc.

File: pageserver/src/tenant/storage_layer/layer.rs (hunk 12 above)
Comment (jcsp) [reply pageserver/src/tenant/storage_layer/layer.rs:1836]: I'm not sure what you're saying here: this method is already on ResidentLayer, so we are already ensuring that the layer is kept resident

File: pageserver/src/tenant/timeline/compaction.rs (hunk 11 above)
Comment (jcsp) [reply pageserver/src/tenant/timeline/compaction.rs]: We don't have access to the layer desc from inside ImageLayerInner::filter.  But we can avoid the range argument by just using a min..max KeySpace and implicitly getting the bounds from the index.  ae5c37b09

File: pageserver/src/tenant/timeline/compaction.rs (hunk 13 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline/compaction.rs:197]: We cannot only use `latest_gc_cutoff_lsn` because that is rarely updated (1/h) compared to pitr_cutoff (6/h). I still think we should hold on to a read guard from `latest_gc_cutoff_lsn` regardless, so that a gc cannot happen while rewriting is ongoing.

File: pageserver/src/tenant/timeline/compaction.rs (hunk 10 above)
Comment (jcsp) [reply pageserver/src/tenant/timeline/compaction.rs]: so in 01504407e, I'm keeping that RcuReadGuard, but afaict that doesn't entirely exclude concurrent GC, it just prevents it entering.  So I've also made rewrite_layers safe to use even if some of the layers were already dropped, by trimming the layers to drop/rewrite _after_ taking the layer map write guard.

Ultimately it probably makes sense to just run not do compaction and GC at the same time, but let's not get into that in this PR

File: pageserver/src/tenant/storage_layer/image_layer.rs (hunk 2 above)
Comment (jcsp) [reply pageserver/src/tenant/storage_layer/image_layer.rs]: This isn't a generic filter function any more.

The plan for a rewrite only becomes large if someone uses a pathologically small stripe size (and users don't have control over that).

File: pageserver/src/tenant/timeline/compaction.rs (hunk 10 above)
Comment (koivunej) [reply pageserver/src/tenant/timeline/compaction.rs]: Yeah I think this PR has a lot of extra stuff already.

File: pageserver/src/tenant/storage_layer/layer.rs (hunk 12 above)
Comment (koivunej) [reply pageserver/src/tenant/storage_layer/layer.rs]: Ok, I got confused by the way DownloadedLayer::get is called. Perhaps it has changed?

File: pageserver/src/tenant/storage_layer/layer.rs [hunk 14]
```diff
@@ -1800,13 +1800,12 @@ impl ResidentLayer {
         use LayerKind::*;
 
         let owner = &self.owner.0;
+        owner
+            .access_stats
+            .record_access(LayerAccessKind::KeyIter, ctx);
```
Comment (koivunej) [inline pageserver/src/tenant/storage_layer/layer.rs:1805]: Why was this moved? It's not a big issue but it was more correct to do it after the kind check.

File: test_runner/regress/test_sharding.py [hunk 15]
```diff
@@ -237,6 +255,90 @@ def test_sharding_split_compaction(neon_env_builder: NeonEnvBuilder):
...
+                "compact-shard-ancestors-localonly",
+                "compact-shard-ancestors-enqueued",
+            ):
+                # If we left local files that don't match remote metadata, we expect warnings on next startup
+                env.pageserver.allowed_errors.append(
+                    ".*removing local file .+ because it has unexpected length.*"
+                )
+
+            # Post-failpoint: we check that the pageserver comes back online happily.
+            env.pageserver.running = False
```
Comment (koivunej) [inline test_runner/regress/test_sharding.py:302]: Is this now needed? I cannot remember setting it ever.

Comment (koivunej) [reply test_runner/regress/test_sharding.py:302]: Aah yes I guess you do it to only start, not restart.

File: pageserver/src/tenant/storage_layer/image_layer.rs [hunk 16]
```diff
@@ -538,6 +553,50 @@ impl ImageLayerInner {
...
+        shard_identity: &ShardIdentity,
+        writer: &mut ImageLayerWriter,
+        ctx: &RequestContext,
+    ) -> anyhow::Result<usize> {
+        // Fragment the range into the regions owned by this ShardIdentity
+        let plan = self
+            .plan_reads(
+                KeySpace {
+                    // If asked for the total key space, plan_reads will give us all the keys in the layer
+                    ranges: vec![Key::MIN..Key::MAX],
```
Comment (koivunej) [inline pageserver/src/tenant/storage_layer/image_layer.rs:569]: This makes sense to me, more than using the layer key range or the caller provided key range.

File: pageserver/src/tenant/storage_layer/image_layer.rs [hunk 17]
```diff
@@ -538,6 +553,50 @@ impl ImageLayerInner {
...
+                ctx,
+            )
+            .await?;
+
+        let vectored_blob_reader = VectoredBlobReader::new(&self.file);
+        let mut key_count = 0;
+        for read in plan.into_iter() {
+            let buf_size = read.size();
+
+            let buf = BytesMut::with_capacity(buf_size);
```
Comment (koivunej) [inline pageserver/src/tenant/storage_layer/image_layer.rs:581]: Why allocate this each loop?

File: pageserver/src/tenant/storage_layer/image_layer.rs [hunk 18]
```diff
@@ -855,3 +914,133 @@ impl Drop for ImageLayerWriter {
...
+        shard::{ShardCount, ShardIdentity, ShardNumber, ShardStripeSize},
+    };
+    use utils::{id::TimelineId, lsn::Lsn};
+
+    use crate::{tenant::harness::TenantHarness, DEFAULT_PG_VERSION};
+
+    use super::ImageLayerWriter;
+
+    #[tokio::test]
+    async fn test_rewrite() -> anyhow::Result<()> {
```
Comment (koivunej) [inline pageserver/src/tenant/storage_layer/image_layer.rs:932]: I do not understand why use `-> anyhow::Result<()>` when it's mixed together with `unwrap`s and `assert!`. There is no upside to using `?` as I understand, it's always a less precise location than a panic would have.

File: pageserver/src/tenant/timeline/compaction.rs [hunk 19]
```diff
@@ -270,25 +278,78 @@ impl Timeline {
...
+        self.rewrite_layers(replace_image_layers, drop_layers)
+            .await?;
+
+        fail::fail_point!("compact-shard-ancestors-enqueued");
 
         // We wait for all uploads to complete before finishing this compaction stage.  This is not
         // necessary for correctness, but it simplifies testing, and avoids proceeding with another
         // Timeline's compaction while this timeline's uploads may be generating lots of disk I/O
         // load.
         self.remote_client.wait_completion().await?;
```
Comment (koivunej) [inline pageserver/src/tenant/timeline/compaction.rs:349]: Doing this unconditionally for sharded at the end of compaction is something new. Note the permit is held at the caller over this.

I do understand why you want it here; recently I added a flag to `timeline_compact` to wait completion altogether, which might be a better play in testing needs vs. real life needs.

**Review by koivunej — APPROVED** [review]

File: pageserver/src/tenant/storage_layer/layer.rs (hunk 14 above)
Comment (jcsp) [reply pageserver/src/tenant/storage_layer/layer.rs:1805]: Moved it back in 010b18aeb -- an earlier version of this change included making load_keys work for image layers..

File: pageserver/src/tenant/storage_layer/image_layer.rs (hunk 17 above)
Comment (jcsp) [reply pageserver/src/tenant/storage_layer/image_layer.rs:581]: This could be optimized, but right now the `read_blobs` interface takes the buffer by move and I'd rather not change it here.

File: pageserver/src/tenant/storage_layer/image_layer.rs (hunk 18 above)
Comment (jcsp) [reply pageserver/src/tenant/storage_layer/image_layer.rs:932]: Changed in 5db0eddf5
